airbyte-integrations/connectors/source-microsoft-onedrive/source_microsoft_onedrive/stream_reader.py

```diff
@@ -95,55 +113,133 @@ def config(self, value: SourceMicrosoftOneDriveSpec):
+                f"Failed to retrieve the initial shared object with ID '{object_id}' from drive '{drive_id}'. "
+                f"HTTP status: {item_response.status_code}. Error: {error_info}"
+            )
+
+        # Check if the object is a file or a folder
+        item_data = item_response.json()
+        if item_data.get("file"):  # Initial object is a file
+            new_path = path + "/" + item_data["name"]
+            last_modified = datetime.strptime(item_data["lastModifiedDateTime"], "%Y-%m-%dT%H:%M:%SZ")
+            yield (new_path, item_data["@microsoft.graph.downloadUrl"], last_modified)
+        else:
+            # Initial object is a folder, start file retrieval
+            yield from get_files(f"{item_url}/children", path)
+
     def list_directories_and_files(self, root_folder, path=None):
         """Enumerates folders and files starting from a root folder."""
         drive_items = root_folder.children.get().execute_query()
         found_items = []
         for item in drive_items:
             item_path = path + "/" + item.name if path else item.name
             if item.is_file:
-                found_items.append((item, item_path))
+                found_items.append((item_path, item.properties["@microsoft.graph.downloadUrl"], item.properties["lastModifiedDateTime"]))
             else:
                 found_items.extend(self.list_directories_and_files(item, item_path))
         return found_items
 
-    def get_files_by_drive_name(self, drives, drive_name, folder_path):
+    def get_files_by_drive_name(self, drive_name, folder_path):
         """Yields files from the specified drive."""
         path_levels = [level for level in folder_path.split("/") if level]
         folder_path = "/".join(path_levels)
 
-        for drive in drives:
-            is_onedrive = drive.drive_type in ["personal", "business"]
-            if drive.name == drive_name and is_onedrive:
+        for drive in self.drives:
+            if drive.name == drive_name:
                 folder = drive.root if folder_path in self.ROOT_PATH else drive.root.get_by_path(folder_path).get().execute_query()
                 yield from self.list_directories_and_files(folder)
 
+    def _get_shared_files_from_all_drives(self):
+        drive_ids = [drive.id for drive in self.drives]
+
+        shared_drive_items = self.one_drive_client.me.drive.shared_with_me().execute_query()
+        for drive_item in shared_drive_items:
+            parent_reference = drive_item.remote_item.parentReference
+
+            # check if drive is already parsed
+            if parent_reference and parent_reference["driveId"] not in drive_ids:
+                yield from self._get_shared_drive_object(parent_reference["driveId"], drive_item.id, drive_item.web_url)
+
+    def get_all_files(self):
+        yield from self.get_files_by_drive_name(self.config.drive_name, self.config.folder_path)
+        yield from self._get_shared_files_from_all_drives()
```

Comment:
How user can choose from which area (his own space or shared) file should be synced?

Reply:
All accessible items will be used, so if the file satisfies pattern it will be synced